File: jar/src/main/java/jline/lang/processes/MarkovChain.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */

package jline.lang.processes;

import jline.util.matrix.Matrix;
import jline.api.mc.Dtmc_stochcomp;

import static jline.api.mc.Dtmc_hitting_time.dtmc_hitting_time;
import static jline.api.mc.Dtmc_isfeasible.dtmc_isfeasible;
import static jline.api.mc.Dtmc_makestochastic.dtmc_makestochastic;
import static jline.api.mc.Dtmc_rand.dtmc_rand;
import static jline.api.mc.Dtmc_solve.dtmc_solve;
import static jline.api.mc.Dtmc_stochcomp.dtmc_stochcomp;
import static jline.api.mc.Dtmc_stochcomp.dtmc_stochcomp_full;
import static jline.api.mc.Dtmc_uniformization.dtmc_uniformization;
import static jline.api.mc.Dtmc_timereverse.dtmc_timereverse;
import static jline.api.mc.Dtmc_transient.dtmc_transient;

/**
 * A class for a discrete time Markov chain
 */
public class MarkovChain extends Process {
    protected Matrix transMat;
    protected Matrix stateSpace;
    protected boolean isfinite;

    /**
     * Creates a DTMC with the specified transition matrix
     * @param transMat the transition matrix
     */
    public MarkovChain(Matrix transMat) {
        this(transMat, true);
    }

    /**
     * Creates a DTMC with the specified transition matrix and finite flag
     * @param transMat the transition matrix
     * @param isFinite whether the DTMC is finite
     */
    public MarkovChain(Matrix transMat, boolean isFinite) {
        super("DTMC", 1);
        this.transMat = dtmc_makestochastic(transMat);
        this.stateSpace = null;
        this.isfinite = isFinite;
    }

    /**
     * Convert to CTMC by subtracting identity matrix
     * @return the equivalent CTMC
     */
    public MarkovProcess toCTMC() {
        Matrix Q = transMat.sub(Matrix.eye(transMat.getNumRows()));
        MarkovProcess ctmc = new MarkovProcess(Q);
        if (stateSpace != null) {
            ctmc.setStateSpace(stateSpace);
        }
        return ctmc;
    }

    /**
     * Get time-reversed DTMC
     * @return time-reversed DTMC
     */
    public MarkovChain toTimeReversed() {
        return new MarkovChain(dtmc_timereverse(transMat));
    }

    /**
     * Get the transition matrix
     * @return the transition matrix
     */
    public Matrix getTransMat() {
        return transMat;
    }

    /**
     * Solve the DTMC for steady-state probabilities. Twin of MarkovProcess.solve.
     * @return the steady-state probability vector
     */
    public Matrix solve() {
        return dtmc_solve(transMat);
    }

    /**
     * Distribution at each step 0,...,steps.
     *
     * @param pi0   initial distribution, uniform when null
     * @param steps number of steps
     * @return matrix with one row per step
     */
    public Matrix transientProb(Matrix pi0, int steps) {
        return dtmc_transient(transMat, pi0, steps);
    }

    /**
     * Mean number of steps to reach any target state.
     *
     * @param targetStates 0-based indices of the target states
     * @return column vector of mean hitting times
     */
    public Matrix hittingTime(int[] targetStates) {
        return dtmc_hitting_time(transMat, targetStates);
    }

    /**
     * Stochastic complement of a subset of states.
     *
     * @param I 0-based indices of the states to retain
     * @return the transition matrix of the complement on those states
     */
    public Matrix stochComp(java.util.List<Integer> I) {
        return dtmc_stochcomp(transMat, I);
    }

    /**
     * Stochastic complement of a subset of states together with the four blocks
     * of the transition matrix partitioned by I and its complement. Twin of the
     * MATLAB [S,P11,P12,P21,P22] = dtmc.stochCompFull(I) and of the Python
     * stochCompFull.
     *
     * @param I 0-based indices of the states to retain
     * @return the complement and the blocks it was built from
     */
    public Dtmc_stochcomp.DtmcStochCompResult stochCompFull(java.util.List<Integer> I) {
        return dtmc_stochcomp_full(transMat, I);
    }

    /**
     * Distribution at time t of the DTMC seen through uniformization. The chain
     * is read as the randomized image of a CTMC, so t is continuous here,
     * unlike the step count taken by transientProb.
     *
     * @param pi0 initial distribution, uniform when null
     * @param t   time point
     * @return the distribution at time t
     */
    public Matrix transientUnif(Matrix pi0, double t) {
        int n = transMat.getNumRows();
        Matrix start = new Matrix(1, n);
        for (int i = 0; i < n; i++) {
            start.set(0, i, (pi0 == null || pi0.length() != n) ? 1.0 / n : pi0.get(i));
        }
        return dtmc_uniformization(start, transMat, t).pi;
    }

    /**
     * @return true when the transition matrix is stochastic
     */
    public boolean isFeasible() {
        return dtmc_isfeasible(transMat) > 0;
    }

    /**
     * Set the state space
     * @param stateSpace the state space matrix
     */
    public void setStateSpace(Matrix stateSpace) {
        this.stateSpace = stateSpace;
    }

    /**
     * Get the state space
     * @return the state space matrix
     */
    public Matrix getStateSpace() {
        return stateSpace;
    }

    /**
     * Check if the DTMC is finite
     * @return true if finite
     */
    public boolean isFinite() {
        return isfinite;
    }

    /**
     * Sample from the DTMC
     * @return sampled value
     */
    @Override
    public Matrix sample() {
        return sample(1);
    }

    /**
     * Sample n steps from the DTMC starting from initial distribution
     * @param n number of steps to sample
     * @return matrix containing sampled states (column vector)
     */
    public Matrix sample(int n) {
        return sample(n, null, new java.util.Random());
    }

    /**
     * Sample n steps from the DTMC
     * @param n number of steps to sample
     * @param pi0 initial state distribution (if null, uses uniform distribution)
     * @param random random number generator
     * @return matrix containing sampled states (column vector)
     */
    public Matrix sample(int n, Matrix pi0, java.util.Random random) {
        int numStates = transMat.getNumRows();
        
        // Set initial distribution if not provided
        if (pi0 == null) {
            pi0 = new Matrix(numStates, 1);
            for (int i = 0; i < numStates; i++) {
                pi0.set(i, 0, 1.0 / numStates);
            }
        }
        
        // Sample initial state
        double r = random.nextDouble();
        double cumulative = 0.0;
        int currentState = 0;
        for (int i = 0; i < numStates; i++) {
            cumulative += pi0.get(i, 0);
            if (r < cumulative) {
                currentState = i;
                break;
            }
        }
        
        Matrix result = new Matrix(n, 1);
        
        // Sample n steps
        for (int step = 0; step < n; step++) {
            result.set(step, 0, currentState);
            
            // Sample next state based on current state's transition probabilities
            if (step < n - 1) { // Don't sample next state for the last step
                r = random.nextDouble();
                cumulative = 0.0;
                for (int j = 0; j < numStates; j++) {
                    cumulative += transMat.get(currentState, j);
                    if (r < cumulative) {
                        currentState = j;
                        break;
                    }
                }
            }
        }
        
        return result;
    }

    /**
     * Create a random DTMC
     * @param nStates number of states
     * @return random DTMC
     */
    public static MarkovChain rand(int nStates) {
        return new MarkovChain(dtmc_rand(nStates));
    }

    /**
     * Create DTMC from sample system aggregation
     * @param samples matrix where each row is a sample trajectory, each column is a time step
     * @return DTMC constructed from samples by estimating transition probabilities
     */
    public static MarkovChain fromSampleSysAggr(Matrix samples) {
        return fromSampleSysAggr(samples, -1);
    }

    /**
     * Create DTMC from sample system aggregation
     * @param samples matrix where each row is a sample trajectory, each column is a time step
     * @param numStates number of states (if -1, infer from data)
     * @return DTMC constructed from samples by estimating transition probabilities
     */
    public static MarkovChain fromSampleSysAggr(Matrix samples, int numStates) {
        if (samples == null || samples.getNumRows() == 0 || samples.getNumCols() < 2) {
            throw new IllegalArgumentException("Invalid sample data: need at least 1 trajectory with 2+ time steps");
        }

        // Determine number of states if not provided
        if (numStates <= 0) {
            double maxState = samples.elementMax();
            double minState = samples.elementMin();
            numStates = (int) (maxState - minState) + 1;
        }

        // Count transitions
        Matrix transitionCounts = new Matrix(numStates, numStates);
        
        for (int trajectory = 0; trajectory < samples.getNumRows(); trajectory++) {
            for (int t = 0; t < samples.getNumCols() - 1; t++) {
                int fromState = (int) samples.get(trajectory, t);
                int toState = (int) samples.get(trajectory, t + 1);
                
                // Validate state indices
                if (fromState >= 0 && fromState < numStates && toState >= 0 && toState < numStates) {
                    transitionCounts.set(fromState, toState, 
                        transitionCounts.get(fromState, toState) + 1.0);
                }
            }
        }

        // Convert counts to probabilities
        Matrix transitionMatrix = new Matrix(numStates, numStates);
        for (int i = 0; i < numStates; i++) {
            double rowSum = 0.0;
            for (int j = 0; j < numStates; j++) {
                rowSum += transitionCounts.get(i, j);
            }
            
            if (rowSum > 0) {
                for (int j = 0; j < numStates; j++) {
                    transitionMatrix.set(i, j, transitionCounts.get(i, j) / rowSum);
                }
            } else {
                // If no transitions observed from state i, make it absorbing
                transitionMatrix.set(i, i, 1.0);
            }
        }

        return new MarkovChain(transitionMatrix);
    }

    /**
     * Create DTMC from sample system aggregation (legacy interface)
     * @param sa the sample aggregation object
     * @return DTMC constructed from samples
     */
    public static MarkovChain fromSampleSysAggr(Object sa) {
        if (sa instanceof Matrix) {
            return fromSampleSysAggr((Matrix) sa);
        } else {
            throw new UnsupportedOperationException("fromSampleSysAggr only supports Matrix input currently");
        }
    }
}

File: matlab/src/lang/processes/MarkovChain.m
classdef MarkovChain < Process
    % An abstract class for a discrete time Markov chain
    %
    % Copyright (c) 2012-2026, Imperial College London
    % All rights reserved.

    properties
        transMat;
        stateSpace;
        isfinite;
    end

    methods
        function self = MarkovChain(transMat, isFinite)
            % SELF = MARKOVCHAIN(transMat, isInfinite)
            self@Process('MarkovChain', 1);

            self.transMat = dtmc_makestochastic(transMat);
            self.stateSpace = [];
            if nargin < 2
                self.isfinite = true;
            else
                self.isfinite = isFinite;
            end
        end

        function A = toMarkovProcess(self)
            Q= self.transMat - eye(size(self.transMat));
            A=MarkovProcess(Q);
            A.setStateSpace(self.stateSpace);
        end
        
        function A = toCTMC(self)
            % TOCTMC - Alias for toMarkovProcess for backwards compatibility
            A = self.toMarkovProcess();
        end

        function Ap = toTimeReversed(self)
            Ap = MarkovChain(dtmc_timereverse(self.transMat));
        end

        function transMat = getTransMat(self)
            transMat = self.transMat;
        end

        function pi = solve(self)
            % PI = SOLVE()
            % Stationary distribution of the DTMC. Twin of MarkovProcess.solve.
            if issym(self.transMat)
                pi = dtmc_solve(self.transMat);
            else
                pi = dtmc_solve_reducible(self.transMat);
            end
        end

        function pi_t = transient(self, pi0, steps)
            % PI_T = TRANSIENT(PI0, STEPS)
            % Distribution at each step 0,...,STEPS from PI0 (uniform if empty).
            if nargin<2
                pi0 = [];
            end
            if nargin<3
                steps = 1;
            end
            pi_t = dtmc_transient(self.transMat, pi0, steps);
        end

        function pi_t = transientProb(self, pi0, steps)
            % PI_T = TRANSIENTPROB(PI0, STEPS)
            % Alias of transient, under the name the JAR must use since
            % 'transient' is a Java keyword.
            if nargin<2
                pi0 = [];
            end
            if nargin<3
                steps = 1;
            end
            pi_t = self.transient(pi0, steps);
        end

        function h = hittingTime(self, targetStates)
            % H = HITTINGTIME(TARGETSTATES)
            % Mean number of steps to reach any state in TARGETSTATES.
            h = dtmc_hitting_time(self.transMat, targetStates);
        end

        function S = stochComp(self, I)
            % S = STOCHCOMP(I)
            % Stochastic complement of the states I, a DTMC on that subset.
            % Use stochCompFull to also obtain the partitioned blocks.
            if nargin<2
                S = dtmc_stochcomp(self.transMat);
            else
                S = dtmc_stochcomp(self.transMat, I);
            end
        end

        function [S, P11, P12, P21, P22] = stochCompFull(self, I)
            % [S, P11, P12, P21, P22] = STOCHCOMPFULL(I)
            % Stochastic complement of the states I together with the blocks of
            % the transition matrix partitioned by I and its complement.
            if nargin<2
                [S, P11, P12, P21, P22] = dtmc_stochcomp(self.transMat);
            else
                [S, P11, P12, P21, P22] = dtmc_stochcomp(self.transMat, I);
            end
        end

        function [pi_t, kmax] = transientUnif(self, pi0, t)
            % [PI_T, KMAX] = TRANSIENTUNIF(PI0, T)
            % Distribution at time T of the DTMC seen through uniformization.
            % The chain is read as the randomized image of a CTMC, so T is
            % continuous here, unlike the step count taken by transient.
            n = size(self.transMat,1);
            if nargin<2 || isempty(pi0)
                pi0 = ones(1,n)/n;
            end
            if nargin<3
                t = 1;
            end
            [pi_t, kmax] = dtmc_uniformization(reshape(pi0,1,[]), self.transMat, t);
        end

        function bool = isFeasible(self)
            % BOOL = ISFEASIBLE()
            % True when the transition matrix is stochastic.
            bool = dtmc_isfeasible(self.transMat) > 0;
        end

        function sts = sample(self, n)
            % STS = SAMPLE(N) - Simulate n steps of the DTMC from a random initial state
            if nargin<2
                n = 1;
            end
            pi0 = rand(1, size(self.transMat,1)); pi0 = pi0/sum(pi0);
            sts = dtmc_simulate(self.transMat, pi0, n);
        end

        function setStateSpace(self,stateSpace)
            self.stateSpace  = stateSpace;
        end

        function plot(self)
            nodeLbl = {};
            if ~isempty(self.stateSpace)
                for s=1:size(self.stateSpace,1)
                    if size(self.stateSpace,2)>1
                        nodeLbl{s} = sprintf('%s%d', sprintf('%d,', self.stateSpace(s,1:end-1)), self.stateSpace(s,end));
                    else
                        nodeLbl{s} = sprintf('%d', self.stateSpace(s,end));
                    end
                end
            end
            P0 = self.transMat;
            [I,J,q]=find(P0);
            edgeLbl = {};
            if ~isempty(self.stateSpace)
                for t=1:length(I)
                    edgeLbl{end+1,1} = nodeLbl{I(t)};
                    edgeLbl{end,2} = nodeLbl{J(t)};
                    edgeLbl{end,3} = sprintf('%.2f',(q(t)));
                end
            else
                for t=1:length(I)
                    edgeLbl{end+1,1} = num2str(I(t));
                    edgeLbl{end,2} = num2str(J(t));
                    edgeLbl{end,3} = sprintf('%.2f',(q(t)));
                end
            end
            if length(nodeLbl) <= 6
                colors = cell(1,length(nodeLbl)); for i=1:length(nodeLbl), colors{i}='w'; end
                graphViz4Matlab('-adjMat',P0,'-nodeColors',colors,'-nodeLabels',nodeLbl,'-edgeLabels',edgeLbl,'-layout',Circularlayout);
            else
                graphViz4Matlab('-adjMat',P0,'-nodeLabels',nodeLbl,'-edgeLabels',edgeLbl,'-layout',Springlayout);
            end
        end

    end

    methods (Static)
        function dtmcObj=rand(nStates) % creates a random DTMC
            dtmcObj = MarkovChain(dtmc_rand(nStates));
        end

        function dtmcObj=fromSampleSysAggr(sa)
            isFinite = true;
            sampleState = sa.state{1};
            for r=2:length(sa.state)
                % per-node trajectories are time-aligned: join column-wise
                sampleState = [sampleState, sa.state{r}];
            end
            [stateSpace,~,stateHash] = unique(sampleState,'rows');
            dtmc = spalloc(length(stateSpace),length(stateSpace),length(stateSpace)); % assume O(n) elements with n states
            holdTime = zeros(length(stateSpace),1);
            for i=2:length(stateHash)
                if isempty(dtmc(stateHash(i-1),stateHash(i)))
                    dtmc(stateHash(i-1),stateHash(i)) = 0;
                end
                dtmc(stateHash(i-1),stateHash(i)) = dtmc(stateHash(i-1),stateHash(i)) + 1;
                holdTime(stateHash(i-1)) = holdTime(stateHash(i-1)) + sa.t(i) - sa.t(i-1);
            end
            % at this point, dtmc has absolute counts so not yet normalized
            dtmc = dtmc_makestochastic(dtmc);
            dtmcObj = MarkovChain(dtmc, isFinite);
            dtmcObj.setStateSpace(stateSpace);
        end
    end
end
